### PR TITLE
Use older Oracle optimizer for faster first exeution 

### PR DESCRIPTION
**Pull request by aidenfoxx**
The standard Oracle cost based optimizer is garbage on data dictionary queries.

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
CI results:

Last run on master:

```
oracledb
[92](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:92)
    .tables
[93](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:93)
      ✓ returns tables (367ms)
[94](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:94)
    .tableInfo
[95](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:95)
      ✓ returns information for all tables
[96](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:96)
      ✓ returns information for specific table (102ms)
[97](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:97)
    .hasTable
[98](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:98)
      ✓ returns if table exists or not (102ms)
[99](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:99)
    .columns
[100](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:100)
      ✓ returns information for all tables (615ms)
[101](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:101)
      ✓ returns information for specific table (49ms)
[102](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:102)
    .columnInfo
[103](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:103)
      ✓ returns information for all columns in all tables (1363ms)
[104](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:104)
      ✓ returns information for all columns in specific table (1068ms)
[105](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:105)
      ✓ returns information for a specific column in a specific table ([106](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:106)9ms)
106
    .hasColumn
[107](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:107)
      ✓ returns has column (88ms)
[108](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:108)
    .primary
[109](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:109)
      ✓ returns primary key for a table (473ms)
[110](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:110)
    .foreignKeys
[111](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:111)
      ✓ returns foreign keys for all tables (231ms)
[112](https://github.com/knex/knex-schema-inspector/runs/5852961767?check_suite_focus=true#step:10:112)
      ✓ filters based on table param (214ms)
```

This PR:

```
oracledb
[113](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:113)
    .tables
[114](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:114)
      ✓ returns tables (180ms)
[115](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:115)
    .tableInfo
[116](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:116)
      ✓ returns information for all tables
[117](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:117)
      ✓ returns information for specific table
[118](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:118)
    .hasTable
[119](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:119)
      ✓ returns if table exists or not
[120](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:120)
    .columns
[121](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:121)
      ✓ returns information for all tables
[122](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:122)
      ✓ returns information for specific table
[123](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:123)
    .columnInfo
[124](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:124)
      ✓ returns information for all columns in all tables (126ms)
[125](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:125)
      ✓ returns information for all columns in specific table (91ms)
[126](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:126)
      ✓ returns information for a specific column in a specific table (86ms)
[127](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:127)
    .hasColumn
[128](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:128)
      ✓ returns has column
[129](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:129)
    .primary
[130](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:130)
      ✓ returns primary key for a table (88ms)
[131](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:131)
    .foreignKeys
[132](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:132)
      ✓ returns foreign keys for all tables (47ms)
[133](https://github.com/knex/knex-schema-inspector/runs/6118452806?check_suite_focus=true#step:10:133)
      ✓ filters based on table param
```

File: lib/dialects/oracledb.ts [hunk 1]
```diff
@@ -244,26 +248,24 @@ export default class oracleDB implements SchemaInspector {
...
+            "rc"."TABLE_NAME" AS "foreign_key_table",
+            "rc"."COLUMN_NAME" AS "foreign_key_column",
+            "uc"."CONSTRAINT_NAME" "constraint_name", 
+            NULL as "on_update", 
+            "uc"."DELETE_RULE" "on_delete" 
+          FROM "USER_CONSTRAINTS" "uc" 
+          INNER JOIN "USER_CONS_COLUMNS" "ucc"
+            ON "uc"."CONSTRAINT_NAME" = "ucc"."CONSTRAINT_NAME" 
+            AND "uc"."CONSTRAINT_TYPE" = 'R'
+          LEFT JOIN "USER_CONS_COLUMNS" "rc"
```
Comment (aidenfoxx) [inline lib/dialects/oracledb.ts:265]:
```suggestion
          INNER JOIN "USER_CONS_COLUMNS" "rc"
```

File: lib/dialects/oracledb.ts [hunk 2]
```diff
@@ -129,69 +130,72 @@ export default class oracleDB implements SchemaInspector {
...
+          FROM 
+            "USER_TAB_COLS" "c" 
+          LEFT JOIN "USER_COL_COMMENTS" "cm"
+            ON "c"."TABLE_NAME" = "cm"."TABLE_NAME" 
+            AND "c"."COLUMN_NAME" = "cm"."COLUMN_NAME" 
+          LEFT JOIN (
+            SELECT /*+ rule */
+              "uc"."TABLE_NAME", 
+              "ucc"."COLUMN_NAME", 
+              "uc"."CONSTRAINT_NAME",
```
Comment (aidenfoxx) [inline lib/dialects/oracledb.ts:159]:
```suggestion
```

**Comment by aidenfoxx** [comment]
Oracle passed. It's just MSSQL being flaky.

**Comment by rijkvanzanten** [comment]
> Oracle passed. It's just MSSQL being flaky.

Seems to be a fluke. Re-running the tests make it pass as expected :|

**Comment by aidenfoxx** [comment]
Updated to use Oracle v11 optimization featurse. Not as good as the rule based optimizer, but still twice as fast on first execution as the current release.

```
  oracledb
    .tables
      ✓ returns tables (239ms)
    .tableInfo
      ✓ returns information for all tables
      ✓ returns information for specific table
    .hasTable
      ✓ returns if table exists or not (56ms)
    .columns
      ✓ returns information for all tables (230ms)
      ✓ returns information for specific table
    .columnInfo
      ✓ returns information for all columns in all tables (486ms)
      ✓ returns information for all columns in specific table (249ms)
      ✓ returns information for a specific column in a specific table (276ms)
    .hasColumn
      ✓ returns has column
    .primary
      ✓ returns primary key for a table (240ms)
    .foreignKeys
      ✓ returns foreign keys for all tables (201ms)
      ✓ filters based on table param (199ms)
```

File: lib/dialects/oracledb.ts [hunk 3]
```diff
@@ -205,7 +232,12 @@ export default class oracleDB implements SchemaInspector {
    */
   async hasColumn(table: string, column: string): Promise<boolean> {
     const result = await this.knex
-      .count<{ count: 0 | 1 }>({ count: '*' })
+      .select<{ count: 0 | 1 }>(
+        this.knex.raw(`
+          /*+ OPTIMIZER_FEATURES_ENABLE('${OPTIMIZER_FEATURES}') NO_QUERY_TRANSFORMATION */
```
Comment (aidenfoxx) [inline lib/dialects/oracledb.ts:237]:
```suggestion
          /*+ OPTIMIZER_FEATURES_ENABLE('${OPTIMIZER_FEATURES}') */
```

File: lib/dialects/oracledb.ts [hunk 4]
```diff
@@ -104,8 +124,11 @@ export default class oracleDB implements SchemaInspector {
   async columns(table?: string) {
     const query = this.knex
       .select<{ table: string; column: string }[]>(
-        'TABLE_NAME as table',
-        'COLUMN_NAME as column'
+        this.knex.raw(`
+          /*+ OPTIMIZER_FEATURES_ENABLE('${OPTIMIZER_FEATURES}') NO_QUERY_TRANSFORMATION */
```
Comment (aidenfoxx) [inline lib/dialects/oracledb.ts:128]:
```suggestion
          /*+ OPTIMIZER_FEATURES_ENABLE('${OPTIMIZER_FEATURES}') */
```

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**